### PR TITLE
Refactoring MQTT client for latest Minimq

### PR DESCRIPTION
**Pull request by ryan-summers**
Follow-on to https://github.com/quartiq/minimq/pull/144

This PR updates the MQTT client to the latest version of Minimq.

### CODE REVIEW

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:286]: @jordens My current thought on this is that we update the `republish_state()` iterator to allow us to simply skip absent topics. Would make this simple, so we can selectively iterate over only the present settings if we want.

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:476]: Current thought is that we just don't cache the properties anymore. Instead, we could only cache:
* The response topic
* The correlation data

Still not entirely sure how to properly cache this data, and I'm not sure if it's required. Perhaps we could just entirely omit the properties when republishing. Alternatively, the user could provide yet another buffer to miniconf to store this cached data into?

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:497]: I think the best thing to do here is to do a proactive check on the `path` before attempting to defer the payload. If the path doesn't exist, then we bail out early. Open to thoughts on this one

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:564]: I think the only time we realistically will cache a response is when we receive a `Get` command. Specifcally because when we receive this command, we do two things:
1. Publish the `path` requested by the `Get` command
2. Reply to the `Get` command

I imagine we could simplify this into a single response somehow, where it's either:
1. Get command succeeded, topic posted to the proper path
2. Get command failed, response provided to the ResponseTopic

With this structure, all requests get exactly 1 reply, which means we don't need to bother with caching anymore.

**Comment by ryan-summers** [comment]
@jordens I'm fighting a few issues currently and wanted to post them here so you could comment on them. I'm curious to hear some of your thoughts. I suspect we may need to add some new `Miniconf` APIs to address some of these changes

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs:497]: The difficulty arises because with the current architecture, we won't realize the path doesn't exist until we go to serialize the publish packet for transmission. I don't know if we can propogate the error all the way back through the stack at that point, so we lose the exact error intel. It could have failed serialization due to a number of other problems, like:
1. No response topic identified
2. Not enough buffer space

File: src/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client.rs:286]: The iterator doesn't know about absent topics. We'd have to extend/change the trait.

What about just erroring from that callback and failing the packet serialization?

File: src/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client.rs:497]: Don't we always send exactly one packet? I.e.

* start serializing the response packet
* as part of that, try the operation (in the callback, late)
* fail operation for reason Err (absent, not found, buffer too small...) or succeed wit Ok
* try serializing the Err/Ok into the response packet
* if that fails, serialize whatever fixed fatal error message can definitely be serialized
* send the packet

File: src/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client.rs:564]: Agreed. That's better. But do send the correlation data both cases.

File: src/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client.rs:497]: ;) Well no, we don't. See [below](https://github.com/quartiq/miniconf/pull/179/files#r1302893601) for that discussion.

Comment (jordens) [reply src/mqtt_client.rs:497]: The question is then can we change the topic (from the get topic to the response topic) while serializing if we hit an err?

File: src/mqtt_client.rs
Comment (jordens) [reply src/mqtt_client.rs:564]: The other option would be to always send the get response to the response topic (Ok or Err) and extend the list command to also (or invent a new command) to be an actual re-publication run (as on startup).

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs]: I have opted to drop the caching of properties, so CD isn't maintained through all the listed elements - just the initial reply to the list request.

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs]: I figured out how to handle this. I updated minimq to allow the error to propagate back from the `publish()` call, so if the serialization fails, we can handle it and retain the error type.

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs]: I updated it to properly handle this condition now.

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs]: Opted for:
1. The `Get` command only sends one reply (always has CD) to the response topic.
2. Failure to serialize in the `Get` command (i.e. path is invalid) will result in a generic failure being sent

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:203]: Gotta figure out the borrowing of the will topic here... The topic doesn't live long enough to just be borrowed

File: src/mqtt_client.rs
Comment (ryan-summers) [inline src/mqtt_client.rs:218]: Not sure what the best choice of action here is if the user doesn't provide us with `will_buffer` memory in the MQTT client config.

Comment (ryan-summers) [reply src/mqtt_client.rs]: Current plan is to refactor the will buffer management.

File: examples/mqtt.rs
Comment (jordens) [inline examples/mqtt.rs:30]: Why does this take a `&mut [u8, N]` and not a `&mut [u8]`?

File: examples/mqtt.rs
Comment (jordens) [inline examples/mqtt.rs:26]: can we try `localhost` (and thus DNS) here?

File: src/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client.rs:217]: Can we leave this for the user to decide?

File: src/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client.rs:220]:
```suggestion
            .autodowngrade_qos()
            .will(will)?;
```

File: src/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client.rs:213]: We should probably either have named them `set_retained()` and `set_qos()` or make them a builder pattern (`let will = Will::new(..).retained(..).qos(..);`)

File: src/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client.rs:314]: Should we not catch `miniconf::Error::Inner()`, i.e. buf too short instead of panicing?

File: src/mqtt_client.rs
Comment (jordens) [inline src/mqtt_client.rs:473]: Could you comment what this returns from? I can't keep the return/continue/break behavior w.r.t match/closures/loops in my head.

File: tests/integration_test.rs
Comment (jordens) [inline tests/integration_test.rs:111]: Ah. Doesn't this clash with the const keepalive in `new()`?

**Review by jordens — CHANGES_REQUESTED** [review]

File: src/lib.rs
Comment (jordens) [inline src/lib.rs:30]: Do we really need this?
Or can we make the part of minimq that miniconf exposes so flexible that using it doesn't require access to additional minimq api to use it? Configbuilder is one of the main blockers here, right?

ref https://github.com/rust-lang/api-guidelines/discussions/176 for some discussion.

File: examples/mqtt.rs
Comment (ryan-summers) [reply examples/mqtt.rs:30]: I'm not sure what you're asking? Do you want to downconvert the buffer using i.e. `&mut buffer[..]`? Not sure what the benefit would be

File: tests/integration_test.rs
Comment (ryan-summers) [reply tests/integration_test.rs:111]: Yeah technically, but they're all 60s. It hink this was just copy-pasta.

File: src/lib.rs
Comment (ryan-summers) [reply src/lib.rs:30]: The minimal API surface would be the `minimq::ConfigBuilder` and all the types that touch it (i.e. buffer config enum, `ProtocolError`, and `Broker` types).

That discussion seems to explicitly recommend a `pub use minimq` here, so I'm a bit confused

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs:306]: It's not clear to me what handling that issue would actually do. I don't think we want to ignore those issues and continue operating, as the user of miniconf would then silently have potential settings that just aren't getting published and/or cannot be set.

What would we do instead of panicking here?

File: examples/mqtt.rs
Comment (jordens) [reply examples/mqtt.rs]: Ah. Here it coerces.
https://github.com/quartiq/minimq/blob/2b466aa1bd7377d59ba58f76172bdd3422d5570c/src/config.rs#L50

File: src/lib.rs
Comment (jordens) [reply src/lib.rs:30]: I think I pasted the wrong link. I has read somewhere else (and that sounds very reasonable to me) that because of possible semver hazards, the re-exports should be absolutely minimal to allow meeting type requirements of the re-exporting crate but to minimize knock-on breaking changes due unrelated stuff changing in the re-exported crate.
Strictly we should only be re-exporting configbuilder, broker, and the relevant error type.
But that's already a lot, hence let's stick with a full re-export.

File: src/mqtt_client.rs
Comment (ryan-summers) [reply src/mqtt_client.rs:307]: If we handle this message and log it, I have concerns. Specifically:
1. Logging is not required on the end-user side, so they may be unaware of the issue
2. A silently unavailable settings seems very problematic

I honestly think an internal panic here is appropriate because it immediately informs the developer at the development stage to fix the issue and doesn't allow it to potentially seep into a production build.

Comment (jordens) [reply src/mqtt_client.rs:307]: I agree that for development a panic is very desirable here.
But for production, the panic is worse than most other options, especially if we know what we can do.
What about instead sending a message to that republication topic, either empty, with some error property attached, or with "<serialization failure>" as payload? Then there is feedback that something is wrong.

Comment (ryan-summers) [reply src/mqtt_client.rs:307]: I think those are both good ideas, I'll take a look at what makes the most sense.

**Comment by ryan-summers** [comment]
@jordens Do we want to wait until minimq releases to land this, or are we fine with leaving a `master` branch dependency until we re-release miniconf?

**Review by jordens — APPROVED** [review]
Looks good!
Let's merge like this, try on stabilizer etc and then do the release ripples.

**Pull request merged**
